xapian-applications/omega/postprocess.py

```diff
@@ -0,0 +1,182 @@
+    ''', formatter_class=argparse.RawTextHelpFormatter)
+parser.add_argument("search_log", type=str, default="log/search.log",
+                    help="Path to the search.log file.")
+parser.add_argument("clicks_log", type=str, default="log/search.log",
+                    help="Path to the clicks.log file.")
+args = parser.parse_args()
+
+
+def generate_final_log(search_log, clicks_log):
+    """Generates Query file formatted as per Xapian Letor documentation.
+
+    Input Args:
+        search_log (str): Path to the search.log file.
+        clicks_log (str): Path to the clicks.log file.
+
+    Example (tab-delimited) entries in search_log:
+
+    821f03288846297c2cf43c34766a38f7	"book"	45,36,14,54,42,52,2,3,15,32	0
+    d41d8cd98f00b204e9800998ecf8427e	""		0
+    d41d8cd98f00b204e9800998ecf8427e	""		0
+    098f6bcd4621d373cade4e832627b4f6	"test"	45,36,14,54,42,52,2,3,15,32	0
+
+    Example (tab-delimited) entries in clicks_log:
+
+    821f03288846297c2cf43c34766a38f7	54
+    821f03288846297c2cf43c34766a38f7	54
+    098f6bcd4621d373cade4e832627b4f6	42
+
+    Example (tab-delimited) entries in final.log:
+
+    QueryID Query   Hits    Offset  Clicks
+    821f03288846297c2cf43c34766a38f7    book    ['45', '36', '14', '54', '42',\
+    '52', '2', '3', '15', '32'] 0   ['45:0', '36:0', '14:0', '54:2', '42:0',\
+    '52:0', '2:0', '3:0', '15:0', '32:0']
+    098f6bcd4621d373cade4e832627b4f6    test    ['45', '36', '14', '54', '42',\
+    '52', '2', '3', '15', '32'] 0   ['45:0', '36:0', '14:0', '54:0', '42:1',\
+    '52:0', '2:0', '3:0', '15:0', '32:0']
+
+    Note: Values in these entries are all of type str.
+    """
+    with open(search_log, 'r') as f, open(clicks_log, 'r') as g:
+        with open('log/final.log', 'w+') as h:
+            reader1 = csv.reader(f, delimiter='\t')
+            reader2 = csv.reader(g, delimiter='\t')
+            writer = csv.writer(h, delimiter='\t')
+
+            # Add headers to final log file
+            writer.writerow(["QueryID", "Query", "Hits", "Offset", "Clicks"])
+
+            qid_to_clicks = collections.defaultdict(dict)
+            did_to_count = {}
+
+            QUERYID, QUERY, HITS = 0, 1, 2
+            DOCID = 1
+
+            # Build map: qid_to_clicks = {qid: {did: click_count}}
+            for row2 in reader2:
+                qid, did = row2[QUERYID], row2[DOCID]
+
+                did_to_count = qid_to_clicks[qid]
+
+                # Check if did is already present in did_to_count
+                if did in did_to_count:
+                    # Update did_to_count[did]
+                    click_count = int(did_to_count[did])
+                    click_count += 1
+                    did_to_count[did] = str(click_count)
+                else:
+                    did_to_count[did] = '1'
+
+                qid_to_clicks[qid] = did_to_count
+
+            f.seek(0)
+
+            clicklist = []
+
+            for row in reader1:
+                # Skip rows with empty Query string
```

Comment:
You could avoid that being written to the log in the first place, maybe. Worth commenting if you're going to keep this in, though.